csrc/scheduler/registry_utils.cpp

```diff
@@ -627,6 +627,42 @@ bool SchedulerTopologyChecker::hasNonNormalizePostReductionBCast(
           continue;
         }
 
+        // Move forward to output to avoid unnecessary segmentation when the
+        // reduction input iteration domain is used after this tensor. See
+        // PostReductionBroadcastCheck test and issue-2146.
+        while (!forward_tv_dep_chain.empty()) {
```

Comment:
This is within another while loop over `forward_tv_dep_chain`, right? Can we simplify the control flow?

Reply:
removed one while loop. The new structure is:
```
 std::vector<std::pair<IterDomain*, IterDomain*>> ids_to_resolve;
 while (!forward_tv_dep_chain.empty()) {
  // Find the id to resolve
  if (ids_to_resolve.empty()) {
    // if stll empty, move forward
    continue;
  }
  // Got at least one id to resolve, try move forward further along the use chain
  if (!forward_tv_dep_chain.empty()) {
    // if can move forward, move forward
    continue;
  }
  // backward resolve

  if (!ids_to_resolve.empty()) {
    // bad case, direct return
    return true;
  }
  // if reach here, ids_to_resolve becomes empty again.
  // start next check along the use chain
 }
```